### PR TITLE
Top devices & top interfaces widgets

### PR DESCRIPTION
**Pull request by spaprzycki**

### CODE REVIEW

File: html/includes/common/top-interfaces.inc.php [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+    $interface_count = $widget_settings['interface_count'];
+    $common_output[] = '
+<h4>Top '.$interface_count.' interfaces (last '.$interval.' minutes)</h4>
+    ';
+    if (is_admin() || is_read()) {
+        $query = '
+            SELECT *, p.ifInOctets_rate + p.ifOutOctets_rate as total
+            FROM ports as p, devices as d
+            WHERE d.device_id = p.device_id
+            AND unix_timestamp() - p.poll_time < '.$interval_seconds.'
```
Comment (f0o) [inline html/includes/common/top-interfaces.inc.php]: use `?` instead of `'.$var.'`

File: html/includes/common/top-interfaces.inc.php [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+      <input class="form-control" name="interface_count" id="input_'.$unique_id.'" placeholder="ie. 5" value="'.$widget_settings['interface_count'].'">
+    </div>
+  </div>
+  <div class="clearfix"></div>
+  <div class="form-group">
+    <div class="col-sm-6">
+      <label for="graph_type" class="control-label">Time interval (minutes): </label>
+    </div>
+    <div class="col-sm-4">
+      <input class="form-control" name="time_interval" id="input_'.$unique_id.'" placeholder="ie. 15" value="'.$widget_settings['time_interval'].'">
```
Comment (f0o) [inline html/includes/common/top-interfaces.inc.php]: you shouldn't have 2x the same ID

File: html/includes/common/top-devices.inc.php [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+      <input class="form-control" name="device_count" id="input_'.$unique_id.'" placeholder="ie. 5" value="'.$widget_settings['device_count'].'">
+    </div>
+  </div>
+  <div class="clearfix"></div>
+  <div class="form-group">
+    <div class="col-sm-6">
+      <label for="graph_type" class="control-label">Time interval (minutes): </label>
+    </div>
+    <div class="col-sm-4">
+      <input class="form-control" name="time_interval" id="input_'.$unique_id.'" placeholder="ie. 15" value="'.$widget_settings['time_interval'].'">
```
Comment (f0o) [inline html/includes/common/top-devices.inc.php]: you shouldn't have 2x the same ID

File: html/includes/common/top-devices.inc.php [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+    $common_output[] = '
+<h4>Top '.$device_count.' devices (last '.$interval.' minutes)</h4>
+    ';
+
+    if (is_admin() || is_read()) {
+        $query = '
+            SELECT *, sum(p.ifInOctets_rate + p.ifOutOctets_rate) as total
+            FROM ports as p, devices as d
+            WHERE d.device_id = p.device_id
+            AND unix_timestamp() - p.poll_time < '.$interval_seconds.'
```
Comment (f0o) [inline html/includes/common/top-devices.inc.php]: use `?` instead of `'.$var.'`

**Comment by f0o** [comment]
The widgets show up empty on the CI, mind double checking?

**Comment by spaprzycki** [comment]
@f0o CI shows widgets empty as there are no devices discovered by CI, so nothing to show in widgets.

**Comment by f0o** [comment]
Correct, I will clean the build something must've gone south with it.
Will report back when home

File: html/includes/common/top-interfaces.inc.php (hunk 2 above)
Comment (spaprzycki) [inline html/includes/common/top-interfaces.inc.php]: Ooops.. sorry! That's fixed now in my local repo, but will defer PR until SQL is sorted - see below.

File: html/includes/common/top-devices.inc.php (hunk 3 above)
Comment (spaprzycki) [inline html/includes/common/top-devices.inc.php]: Ooops.. sorry! That's fixed now in my local repo, but will defer PR until SQL is sorted - see below.

File: html/includes/common/top-devices.inc.php (hunk 4 above)
Comment (spaprzycki) [inline html/includes/common/top-devices.inc.php]: That's how I wanted to do it initially, but could not figure out a way (or find example) how to do it with different number of parameters depending on which query got selected. Non admin query has 3 parameters while admin query has only 2. The extra parameter comes first in non-admin query, so how do I 'tell' dbFetch to skip it on admin query? Do I need two dbFetch calls?

I was looking around repo to find how other did that in the past, but failed to find one - all either have single parameter or use '.$var.' notation.

Comment (f0o) [inline html/includes/common/top-devices.inc.php]: I'll paste you a patch for it when home

**Comment by f0o** [comment]
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

New build looks good from what I could see via mobile. :+1:

I'll be back in about 2 hrs max, would be nice to catch up on irc and get this merged :)

On September 13, 2015 3:51:58 PM GMT+02:00, LibreNMS-CI <notifications@github.com> wrote:
>Auto-Deploy finished, Test PR at http://1894.ci.librenms.org or
>https://1894.ci.librenms.org
>
>
>---
>Reply to this email directly or view it on GitHub:
>https://github.com/librenms/librenms/pull/1894#issuecomment-139878351
-----BEGIN PGP SIGNATURE-----
Version: APG v1.1.1

iQJPBAEBCgA5BQJV9YLTMhxEYW5pZWwgUHJldXNza2VyIChNb2JpbGVLZXkjMikg
PGYwb0BsaWJyZW5tcy5vcmc+AAoJEJMZKtVWOFzB0NUP/3E/p8x9JKl74ZdVM/LC
6QGSrdA5dXeKGjy7iYzSVD7yFT4QH5B7cITZ+efCQ7/z/jmLVGE688QNRqS+3xKU
oKnx+BikvQVtySF5SqkRLSjLmZJjZJD+GSgZU/hFOZ/chWyXhZJ3z8DpnrG7VJzz
AYYR1nXuzIHljQtXM/q6nmDn8SFMSlNjLSd5EGHjiC1Q1WSVLiGLVl5qqrxgylzI
ARJrln7DERZ2MxMplOh/Et3cQheTcIQsg+1JteEw3BZ2MbIt/2YBX7GRJYFjyUAi
elUWAd/Ijctfhp/F6oeEc9F99LIOroCDbPp5DgBVafrZ/NxK7UYlo6oyCvVKb0sa
obDornZryqTPRIcjbAUSuziCRFQPvN8/mhPhrGDhCq2oSCT4X9qFm3cUcr4rkHzC
xTlqbDDdg49kK5OTmExOUeuv2a9Ikuii7NTEYgSnxVHoRL4SAfknKU1mAANDDgPh
4ZQdPIZJoQTipLGv1nZgmiRhEZrqUfSx145c4vLhC+LNLBvuynne/F9n4wuj1gYD
p5hrlrEaWjPeK9tB/H0646+QnbjrFwf7BAkknVBcLNovntZAIjk0+7jv10/rK5D5
HqEx26gHzOCgH/8aUswDoy0y64jBE3V1D+tOBr1Sn7V5X3/2NArKpHwKURMLhRkt
4JYmXdD7EYHpMg540NaQ0G3E
=khC1
-----END PGP SIGNATURE-----

**Comment by spaprzycki** [comment]
@f0o I'll try to be around as much as wife and kids let me be ;-)

File: html/includes/common/top-interfaces.inc.php [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+            OR (`PP`.`user_id` = ? AND `PP`.`port_id` = `I`.`port_id` 
+            AND `I`.`device_id` = `d`.`device_id`)) AND
+            d.device_id = I.device_id
+            AND unix_timestamp() - I.poll_time < ? 
+            AND ( I.ifInOctets_rate > 0
+            OR I.ifOutOctets_rate > 0 )
+            ORDER BY total desc
+            LIMIT ?
+        ';
+        $params = array($_SESSION['user_id'], $interval_seconds, $device_count);
```
Comment (f0o) [inline html/includes/common/top-interfaces.inc.php]: `$_SESSION['user_id']` has to be passed twice as the query uses it twice.

File: html/includes/common/top-devices.inc.php [hunk 6]
```diff
@@ -0,0 +1,125 @@
...
+            WHERE `P`.`user_id` = ? AND `P`.`device_id` = `d`.`device_id` AND
+            d.device_id = p.device_id
+            AND unix_timestamp() - p.poll_time < ? 
+            AND ( p.ifInOctets_rate > 0
+            OR p.ifOutOctets_rate > 0 )
+            GROUP BY d.device_id
+            ORDER BY total desc
+            LIMIT ?
+            ';
+        $params = array($_SESSION['user_id'], $interval_seconds, $device_count);
```
Comment (f0o) [inline html/includes/common/top-devices.inc.php]: See below

**Comment by spaprzycki** [comment]
@f0o - it seems to be working well now. sorry for mess with commits, but I'm still learning and it still goes out of control when I do something what I shouldn't have done ;-)

**Comment by f0o** [comment]
Closing in favour of the rebased PR #1903

**Pull request closed without merging**